File: algos/leetcode/numberofislands.py
#https://leetcode.com/problems/number-of-islands/
class Solution:

    def dfs(self, i, j, grid:  List[List[str]]):

        n = len(grid[0])
        m = len(grid)

        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] == '0':
            return
        
        grid[i][j] = '0'

        self.dfs(i + 1, j, grid)
        self.dfs(i - 1, j, grid)
        self.dfs(i, j + 1, grid)
        self.dfs(i, j - 1, grid)

    def numIslands(self, grid: List[List[str]]) -> int:

        n = len(grid[0])
        m = len(grid)

        if m == 0 or n == 0:
            return 0
                
        numberOfIslands = 0
        for i in range(0, len(grid)):
            for j in range(0, len(grid[0])):
                if grid[i][j] == '1':
                    numberOfIslands +=1
                    self.dfs(i , j , grid)

        return numberOfIslands 

File: algos/leetcode/number-of-islands.java
#https://leetcode.com/problems/number-of-islands/submissions/
class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0;
        }
        
        int m = grid.length;
        int n = grid[0].length;
        int numIslands = 0;
        
        // Iterate through each cell in the grid
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    // Found an unvisited island, perform DFS to mark all connected land cells
                    numIslands++;
                    dfs(grid, i, j);
                }
            }
        }
        
        return numIslands;
    }
    
    private void dfs(char[][] grid, int i, int j) {
        int m = grid.length;
        int n = grid[0].length;
        
        // Boundary check and check if cell is land ('1')
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == '0') {
            return;
        }
        
        // Mark the current cell as visited by changing '1' to '0'
        grid[i][j] = '0';
        
        // Recursively visit all adjacent land cells
        dfs(grid, i + 1, j); // Down
        dfs(grid, i - 1, j); // Up
        dfs(grid, i, j + 1); // Right
        dfs(grid, i, j - 1); // Left
    }
}
